runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperatorTest.java

```diff
@@ -305,7 +305,10 @@ public void processElement(
             eventTimerWithOutputTimestamp
                 .withOutputTimestamp(timerOutputTimestamp)
                 .set(timerTimestamp);
-            processingTimer.offset(Duration.millis(timerTimestamp.getMillis())).setRelative();
+            processingTimer
```

Comment:
It seems to me that the test with regard to the watermark contract makes little sense for processing time timers. If we want to keep the test, can we change the validation to verify that the output timestamp is equal to output watermark instead of explicitly setting the output timestamp?

I must say I'm not sure how to interpret user defined `withOutputTimestamp` for processing time timers (and even less for relative timers as this one). There is no relation between the two time domains and processing time timers do not hold output watermark, that seems to be very error prone. cc @reuvenlax

Reply:
> processing time timers do not hold output watermark

Possibly not all runners implement it, but certainly on Dataflow processing time timers hold the watermark at their output timestamp.  And again, if the user doesn't explicitly set the output timestamp, it is defaulted to the timestamp of the element (or timer) that set it.

Reply:
OK, I somehow managed overlook this. :)

Two notes:
 a) yes, for instance FlinkRunner definitely holds watermarks only for event-time timers (this is bad, looks like missing more @ValidatesRunner tests) and
 b) this seems inconsistent with how event-time timers work, which set output timestamp based on the fire timestamp (by default). Do we have a reason for such a difference? And what about looping processing-time timers, where there is actually no element? The default of output being the output watermark seems logically consistent with the event-time timers, because the output watermark gives the event-time position.

Reply:
> this seems inconsistent with how event-time timers work, which set output timestamp based on the fire timestamp (by default). Do we have a reason for such a difference?

My guess would be that "firing timestamp" is in a different time domain here, so element input timestamp is the only thing that can be used that's in the event time domain to set the hold on.

> And what about looping processing-time timers, where there is actually no element?

The output timestamp of the timer (the timer executing the OnTimer callback) is [used](https://github.com/apache/beam/blob/master/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L674).

I actually do agree that the default being the output watermark if the output timestamp on the timer is unset makes more sense.  This would also remove the need for a watermark hold at all, and remove a source of confusion around processing time timers.  It seems like it'd be a larger change though.

Reply:
It seems that this change was actually introduced later. Which is why the comment on `setAndVerifyOutputTimestamp` says that processing-time timers should not have output timestamp.

> My guess would be that "firing timestamp" is in a different time domain here, so element input timestamp is the only thing that can be used that's in the event time domain to set the hold on.

The different time domain relates to *when* the timer should fire. The default output timestamp can be (and I think it should be) consistent in both cases - it is the position in *event-time* in both cases. In the case of event-time timer that is the firing timestamp, in the case of processing-time timer it is the output watermark.

Agree that the change to make this consistent would be a larger work, but we should know what the correct behavior should be and if we actually have runners that violate the model.

Reply:
to close the loop here, I updated the Flink DoFnOperator to preserve the existing behavior.  In a follow up review we can fix it to set a watermark hold on the output timestamp of processing time timers (essentially unifying them with event time timers).

Reply:
My bad. This fixes it for both code paths. OK, makes sense. :+1:

Reply:
+1 on their being two different timestamps, the firing timestamp and the output timestamp.

The firing timestamp can be in the processing time domain or the event time domain.
The output timestamp is always in the event time domain and is used to hold the watermark.

For event time timers, the output timestamp being equivalent to the firing time was expected to be a natural default since users would want to produce data with the same output timestamp as the firing timestamp but users could still override it by setting the output timestamp separately from the firing timestamp.

For processing time timers, the output timestamp can't rely on the firing time as it doesn't make sense as a default since it represents a different time domain. This is why the output timestamp defaults to the input element/timer timestamp and can be overridden to a different output timestamp.

Also, any scheduled timers should have an input timestamp equal to the output timestamp that was set allowing for new timers that are scheduled from the timer callback to get reasonable defaults as well.